drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/client/DriftNettyMethodInvokerFactory.java

```diff
@@ -66,65 +65,74 @@ public DriftNettyMethodInvokerFactory(DriftNettyConnectionFactoryConfig factoryC
-        this.defaultSocksProxy = factoryConfig.getSocksProxy();
+        this.defaultSocksProxy = Optional.ofNullable(factoryConfig.getSocksProxy());
+
+        ConnectionManager connectionManager = new ConnectionFactory(group, sslContextFactory);
+        if (factoryConfig.isConnectionPoolEnabled()) {
+            connectionManager = new ConnectionPool(connectionManager, group, factoryConfig.getConnectionPoolMaxSize(), factoryConfig.getConnectionPoolIdleTimeout());
+        }
+        this.connectionManager = connectionManager;
     }
 
     @Override
     public MethodInvoker createMethodInvoker(I clientIdentity)
     {
-        DriftNettyClientConfig clientConfig = clientConfigurationProvider.apply(clientIdentity);
+        ConnectionParameters clientConfig = toConnectionConfig(clientConfigurationProvider.apply(clientIdentity));
+
+        // validate ssl context configuration is valid
+        clientConfig.getSslContextConfig()
+                .ifPresent(sslContextConfig -> sslContextFactory.get(sslContextConfig).get());
+
+        return new DriftNettyMethodInvoker(clientConfig, connectionManager, group);
+    }
+
+    @PreDestroy
+    @Override
+    public void close()
+    {
+        try {
+            connectionManager.close();
+        }
+        finally {
+            try {
+                group.shutdownGracefully(0, 0, MILLISECONDS).await();
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    private ConnectionParameters toConnectionConfig(DriftNettyClientConfig clientConfig)
+    {
         if (clientConfig.getTransport() == HEADER && clientConfig.getProtocol() == COMPACT) {
             throw new IllegalArgumentException("HEADER transport cannot be used with COMPACT protocol, use FB_COMPACT instead");
         }
-        if (clientConfig.getSocksProxy() == null) {
-            clientConfig.setSocksProxy(defaultSocksProxy);
-        }
 
-        Optional<Supplier<SslContext>> sslContext = Optional.empty();
+        Optional<SslContextConfig> sslContextConfig = Optional.empty();
         if (clientConfig.isSslEnabled()) {
-            sslContext = Optional.of(sslContextFactory.get(
+            sslContextConfig = Optional.of(new SslContextConfig(
                     clientConfig.getTrustCertificate(),
                     Optional.ofNullable(clientConfig.getKey()),
                     Optional.ofNullable(clientConfig.getKey()),
                     Optional.ofNullable(clientConfig.getKeyPassword()),
                     clientConfig.getSessionCacheSize(),
                     clientConfig.getSessionTimeout(),
                     clientConfig.getCiphers()));
+        }
 
-            // validate ssl context configuration is valid
-            sslContext.get().get();
+        Optional<HostAndPort> socksProxy = Optional.ofNullable(clientConfig.getSocksProxy());
+        if (!socksProxy.isPresent()) {
```

Comment:
I thought we could do better here, but it needs the new `or()` method from Java 9